### PR TITLE
Introduce time-based retention policy for soft-deletes

### PR DESCRIPTION
**Pull request by dnhatn**
Today soft-deleted documents retention policy is configured by an
absolute number of operations before the global checkpoint. Having 
the right default setting for this policy is hard.

This change introduces a time-based retention policy. I see two possible
ways to implement a time-based policy:

1. Use an indexed field. While merging might be faster with this
approach, we have to add an extra indexed field for every document (an
optimization is to reuse a user-defined timestamp field).

2. Use a numeric docValues field. While the retention query of this approach
might be slower than the previous's, but we only have to add a docValues 
to the updated (i.e, soft-deleted) documents and tombstones.

This PR chooses the second approach.

Relates #34908

/cc @martijnvg and @jasontedor

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/SoftDeletesPolicy.java [hunk 1]
```diff
@@ -114,7 +127,20 @@ synchronized long getMinRetainedSeqNo() {
...
      */
-    Query getRetentionQuery() {
-        return LongPoint.newRangeQuery(SeqNoFieldMapper.NAME, getMinRetainedSeqNo(), Long.MAX_VALUE);
+    synchronized Query getRetentionQuery() {
+        final BooleanQuery.Builder queryBuilder = new BooleanQuery.Builder();
+        queryBuilder.add(LongPoint.newRangeQuery(SeqNoFieldMapper.NAME, getMinRetainedSeqNo(), Long.MAX_VALUE), BooleanClause.Occur.SHOULD);
+        if (retentionAge.seconds() > 0) {
+            // we prefer using a docValues instead of an indexed field to avoid impact on append-only indices.
+            // however, it's a trade-off as we have to linearly scan to find the most recent documents.
+            final Query timestampQuery = NumericDocValuesField.newSlowRangeQuery(SeqNoFieldMapper.UPDATE_TIMESTAMP_IN_SECONDS_NAME,
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/engine/SoftDeletesPolicy.java]: I think this the right trade off here.

Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/engine/SoftDeletesPolicy.java]: Can you explain that a bit more? If there are no soft deleted documents in a segment, the overhead of an indexed field is 0? also, the overhead is linear in the number of soft-deleted documents? The upside of an indexed field is that it's much faster to query to see if we need to merge a segment away.

Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/SoftDeletesPolicy.java]: @bleskes I added a comment.

Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/engine/SoftDeletesPolicy.java]: Thanks @dnhatn. From you comment I read that the main point here is that we need to update existing docs with this value (with soft deletes) or store it there originally  (with an index field).

Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/SoftDeletesPolicy.java]: @bleskes Good point. I'll add it.

**Comment by s1monw** [comment]
ok so this is a rough idea but maybe it triggers some follow ups. What if we instead of adding this to the index keep a datastructure in memory that holds an upper bound of sequence IDs for every 10 min into the past and reduce granularity the further we go into the future. If we can keep this rather smallish datastructure within the bounds of a couple of numbers we can just put it into the commit metadata and don't need any additional fields and the query will be a very simple one?

**Comment by dnhatn** [comment]
@s1monw This approach eliminates the extra query and extra docValues; and also allows us to use a time-based policy in peer-recovery which is not the case with the current implementation. I will explore this a bit more and hopefully come with a prototype soon.

**Comment by s1monw** [comment]
I will start looking into this @jasontedor / @bleskes

**Comment by s1monw** [comment]
@dnhatn should we close this for now?

**Comment by dnhatn** [comment]
@s1monw Yes. I am closing this in favor of https://github.com/elastic/elasticsearch/issues/37165.

**Pull request closed without merging**